File: cpp/tests/BinaryTreePreorderTraversalTest.cpp
#include "BinaryTreePreorderTraversal.h"
#include "gtest/gtest.h"

using namespace lcpp;

TEST(BinaryTreePreorderTraversal, Solution144_1) {
  auto S144_1 = Solution144_1();

  auto Root1 = newTree({1, NTNode, 2, 3});
  std::vector<int> const Expected1{1, 2, 3};
  EXPECT_EQ(Expected1, S144_1.preorderTraversal(Root1));
  deleteTree(Root1);

  auto Root2 = newTree({});
  std::vector<int> const Expected2;
  EXPECT_EQ(Expected2, S144_1.preorderTraversal(Root2));
  deleteTree(Root2);

  auto Root3 = newTree({1});
  std::vector<int> const Expected3{1};
  EXPECT_EQ(Expected3, S144_1.preorderTraversal(Root3));
  deleteTree(Root3);
}

TEST(BinaryTreePreorderTraversal, Solution144_2) {
  auto S144_2 = Solution144_2();

  auto Root1 = newTree({1, NTNode, 2, 3});
  std::vector<int> const Expected1{1, 2, 3};
  EXPECT_EQ(Expected1, S144_2.preorderTraversal(Root1));
  deleteTree(Root1);

  auto Root2 = newTree({});
  std::vector<int> const Expected2;
  EXPECT_EQ(Expected2, S144_2.preorderTraversal(Root2));
  deleteTree(Root2);

  auto Root3 = newTree({1});
  std::vector<int> const Expected3{1};
  EXPECT_EQ(Expected3, S144_2.preorderTraversal(Root3));
  deleteTree(Root3);
}

File: java/src/test/java/io/binac/leetcode/BinaryTreePreorderTraversalTest.java
package io.binac.leetcode;

import io.binac.leetcode.util.TreeNode;
import io.binac.leetcode.util.Trees;
import org.junit.jupiter.api.Test;

import java.util.Arrays;
import java.util.Collections;
import java.util.List;

import static org.junit.jupiter.api.Assertions.*;

class BinaryTreePreorderTraversalTest {
    private static final BinaryTreePreorderTraversal.Solution1 SOLUTION1 = new BinaryTreePreorderTraversal.Solution1();
    private static final BinaryTreePreorderTraversal.Solution2 SOLUTION2 = new BinaryTreePreorderTraversal.Solution2();

    private final TreeNode ROOT1 = Trees.asTree(1, null, 2, 3);
    private final List<Integer> EXPECTED1 = Arrays.asList(1, 2, 3);

    private final TreeNode ROOT2 = Trees.asTree();
    private final List<Integer> EXPECTED2 = Collections.emptyList();

    private final TreeNode ROOT3 = Trees.asTree(1);
    private final List<Integer> EXPECTED3 = Collections.singletonList(1);

    @Test
    void test1() {
        assertEquals(EXPECTED1, SOLUTION1.preorderTraversal(ROOT1));
        assertEquals(EXPECTED2, SOLUTION1.preorderTraversal(ROOT2));
        assertEquals(EXPECTED3, SOLUTION1.preorderTraversal(ROOT3));
    }

    @Test
    void test2() {
        assertEquals(EXPECTED1, SOLUTION2.preorderTraversal(ROOT1));
        assertEquals(EXPECTED2, SOLUTION2.preorderTraversal(ROOT2));
        assertEquals(EXPECTED3, SOLUTION2.preorderTraversal(ROOT3));
    }
}